File: Day_4/almost_sorted.js
// Complete the almostSorted function below.
function almostSorted(arr) {
    var tempArr1 = arr.slice();
    arr.sort((a, b) => { return a - b; });
    var tempArr2 = [];

    for (let i = 0; i < arr.length; i++){
        if (arr[i] != tempArr1[i]) {
            tempArr2.push(i);
        }
    }
    if (tempArr1.length === 2) {
        console.log("yes");
        console.log("swap " + (tempArr2[0] + 1) + " " + (tempArr2[1] + 1));
    }
    else if (tempArr2.length === 0) {
        console.log("yes");
    }
    else {
        var l = tempArr2[0], r = tempArr2[tempArr2.length - 1];
        while (l < r) {
            var temp = tempArr1[l];
            tempArr1[l] = tempArr1[r];
            tempArr1[r] = temp;
            l++;
            r--;
        }
    }
    var h = [];
    for (let i = 0; i < arr.length; i++){
        if (arr[i] != tempArr1[i]) {
            h.push(i);
        }
    }
    if (h.length === 0) {
        console.log("yes");
        console.log("reverse " + (tempArr2[0] + 1) + " " + (tempArr2[tempArr2.length - 1] + 1));
    }
    else if (h.length > 0 && tempArr2.length > 2) {
        console.log("no");
    }
}

File: Day_4/almost_sorted.java
 static void almostSorted(int[] arr) {
        int n = arr.length;
        int[] sortedArray = new int[n];
        for(int i = 0; i < n; i++){
            sortedArray[i] = arr[i];
        }
        Arrays.sort(sortedArray);
        ArrayList<Integer> difference = new ArrayList<Integer>();
        for(int i = 0; i < n; i++) {
            if(arr[i] != sortedArray[i]) {
                difference.add(i);
                if(difference.size() == 3) {
                    break;
                }
            }
        }
        if(difference.size() == 2) {
            System.out.println("yes\nswap " + (difference.get(0) + 1) + " " + (difference.get(1) + 1));
        } else {
            int first = -1;
            int last = Integer.MAX_VALUE;
            for(int i = 0; i < n; i++) {
                if(arr[i] != sortedArray[i]) {
                    if(first == -1) {
                        first = i;
                    }
                    last = i;
                }
            }
            if(first == -1) {
                System.out.println("no");
            } else {
                boolean works = true;
                for(int i = 0; i < last - first; i++) {
                    if(arr[first + i] != sortedArray[last - i]) {
                        works = false;
                        break;
                    }
                }
                if(works) {
                    System.out.println("yes\nreverse " + (first + 1) + " " + (last + 1));
                } else {
                    System.out.println("no");
                }
            }
        }
    }